src/domains/list_iterator.h

```diff
@@ -0,0 +1,77 @@
+/*******************************************************************\
+
+Module: List iterator - abstraction for iterative access to a linked
+                        list.
+
+Author: Viktor Malik
+
+\*******************************************************************/
+#ifndef CPROVER_2LS_DOMAINS_LIST_ITERATOR_H
+#define CPROVER_2LS_DOMAINS_LIST_ITERATOR_H
+
+
+#include <util/std_expr.h>
+#include <set>
+
+class list_iteratort
+{
+public:
+  // No location (used for input variables)
+  static const int IN_LOC=-1;
+
+  /*******************************************************************\
+   Access to an object from a list iterator.
+   Contains:
+     - sequence of fields that are used to access the object from the
+       iterator
+     - location:
+         IN_LOC for read access
+         location number for write access
+  \*******************************************************************/
+  class accesst
+  {
+  public:
+    std::vector<irep_idt> fields;
+    int location;
+
+    equal_exprt binding(
+      const symbol_exprt &lhs, const symbol_exprt &rhs,
+      const unsigned level, const namespacet &ns) const;
+  };
+
+  // Pointer variable used to iterate the list (induction pointer)
+  symbol_exprt pointer;
+  // Initial value of the induction pointer (before the first iteration)
+  exprt init_pointer;
+  // Set of fields through which a step is done after each iteration
+  std::vector<irep_idt> fields;
+  mutable std::list<accesst> accesses;
```

Comment:
why is this mutable? Rather make the caller non-const.

Reply:
because we want to add accesses to an iterator that is already in a set